In [276]:
import pandas as pd
import numpy as np

In [277]:
train=pd.read_csv("MIP_train_final.csv")
test=pd.read_csv("MIP_test_final.csv")
voting=pd.read_csv("MIP Voting.csv")

In [278]:
try:
    train=train.drop(columns=["MP.2","MP.3"])
    #train=train.drop(columns=["MP.2_diff","MP.3_diff"])
    test=test.drop(columns=["MP.2","MP.3"])
    test=test.drop(columns=["MP.2_diff","MP.3_diff"])
except:
    pass

In [279]:
train.head(20)

br_id  season  age team league pos  games  Games Started    MP  \
0   roberal01    1985   22  SAS    NBA  SG     79              9  21.3   
1   roberal01    1986   23  SAS    NBA  SG     82             82  35.1   
2   barklch01    1985   21  PHI    NBA  PF     82             60  28.6   
3   barklch01    1986   22  PHI    NBA  PF     80             80  36.9   
4   willike02    1985   22  ATL    NBA  PF     82             19  21.8   
5   willike02    1986   23  ATL    NBA  PF     82             59  28.0   
6   willike02    1991   28  ATL    NBA  PF     80             80  29.7   
7   willike02    1992   29  ATL    NBA  PF     81             80  36.6   
8   wilkido01    1985   25  ATL    NBA  SF     81             81  37.3   
9   wilkido01    1986   26  ATL    NBA  SF     78             78  39.1   
10  duckwke01    1987   22  POR    NBA   C     51              0  14.8   
11  duckwke01    1988   23  POR    NBA   C     78             50  28.5   
12  stockjo01    1987   24  UTA    NBA  PG     82              2  22.7   
13  stockjo01    1988   25  UTA    NBA  PG     82             79  34.7   
14  pricema01    1987   22  CLE    NBA  PG     67              0  18.2   
15  pricema01    1988   23  CLE    NBA  PG     80             79  32.8   
16  adamsmi01    1987   24  WSB    NBA  PG     63              0  20.7   
17  adamsmi01    1988   25  DEN    NBA  PG     82             75  33.9   
18  adamsmi01    1990   27  DEN    NBA  PG     79             74  34.1   
19  adamsmi01    1991   28  DEN    NBA  PG     66             66  35.5   

    MP Total  ...  TOV%  USG%  OWS  DWS    WS  WS/48  OBPM  DBPM  BPM  VORP  
0       1685  ...  19.8  20.0  1.1  2.0   3.1  0.087  -0.2   1.1  0.9   1.2  
1       2878  ...  17.1  20.9  4.0  4.6   8.6  0.143   1.5   3.3  4.7   4.9  
2       2347  ...  17.9  20.0  4.2  3.3   7.5  0.154   1.3   0.4  1.7   2.2  
3       2952  ...  21.3  22.4  5.4  5.5  10.8  0.176   3.1   2.1  5.1   5.3  
4       1785  ...  11.9  20.2  0.4  2.2   2.7  0.072  -2.3  -1.4 -3.8  -0.8  
5       2300  ...  16.0  19.6  1.4  3.3   4.7  0.099  -1.3  -0.6 -1.9   0.0  
6       2373  ...  13.4  19.8  2.3  2.0   4.4  0.089  -0.3  -1.5 -1.8   0.1  
7       2962  ...  12.5  22.0  4.7  4.1   8.8  0.143   1.6  -1.4  0.2   1.7  
8       3023  ...   9.4  32.5  4.1  3.5   7.6  0.121   3.1  -0.9  2.2   3.2  
9       3049  ...  10.2  32.9  7.0  3.9  10.8  0.170   4.6  -0.5  4.2   4.8  
10       753  ...  20.9  18.5 -0.1  0.8   0.7  0.043  -5.3  -0.9 -6.2  -0.8  
11      2223  ...  13.9  22.6  3.0  2.1   5.2  0.112  -0.8  -2.0 -2.8  -0.4  
12      1858  ...  22.5  15.5  3.4  3.3   6.7  0.174   1.6   3.7  5.3   3.4  
13      2842  ...  21.9  17.2  9.2  4.9  14.1  0.238   5.2   3.3  8.5   7.6  
14      1217  ...  18.1  19.1  0.0  1.1   1.1  0.043  -1.7  -0.9 -2.6  -0.2  
15      2626  ...  14.5  20.6  5.6  2.2   7.8  0.143   2.9  -0.7  2.2   2.8  
16      1303  ...  15.3  16.3  1.2  1.5   2.7  0.099   0.0   1.2  1.2   1.0  
17      2778  ...  12.4  16.2  4.4  3.2   7.5  0.130   1.5   0.9  2.5   3.1  
18      2690  ...  11.1  18.5  4.4  2.5   6.9  0.124   1.8  -0.1  1.8   2.6  
19      2346  ...  12.7  28.5  5.8  0.4   6.3  0.128   6.0  -0.7  5.3   4.3  

[20 rows x 52 columns]

In [280]:
voting

Rank                   Player  Age   Tm  Year   Season  First  Pts Won  \
0      1          Alvin Robertson   23  SAS  1986  1985-86   30.0     30.0   
1      2          Charles Barkley   22  PHI  1986  1985-86   18.5     18.5   
2      3             Kevin Willis   23  ATL  1986  1985-86    8.0      8.0   
3      4        Dominique Wilkins   26  ATL  1986  1985-86    6.5      6.5   
4      1          Kevin Duckworth   23  POR  1988  1987-88   33.0     33.0   
..   ...                      ...  ...  ...   ...      ...    ...      ...   
760  19T  Shai Gilgeous-Alexander   22  OKC  2021  2020-21    0.0      1.0   
761  19T           Richaun Holmes   27  SAC  2021  2020-21    0.0      1.0   
762  19T           T.J. McConnell   28  IND  2021  2020-21    0.0      1.0   
763  19T             Terry Rozier   26  CHO  2021  2020-21    0.0      1.0   
764  19T           Andrew Wiggins   25  GSW  2021  2020-21    0.0      1.0   

     Pts Max  Share  ...   TRB  AST  STL  BLK    FG%    3P%    FT%    WS  \
0         78  0.385  ...   6.3  5.5  3.7  0.5  0.514  0.276  0.795   8.6   
1         78  0.237  ...  12.8  3.9  2.2  1.6  0.572  0.227  0.685  10.8   
2         78  0.103  ...   8.6  0.5  0.8  0.5  0.517  0.000  0.654   4.7   
3         78  0.083  ...   7.9  2.6  1.8  0.6  0.468  0.186  0.818  10.8   
4         80  0.413  ...   7.4  0.8  0.4  0.4  0.496    NaN  0.770   5.2   
..       ...    ...  ...   ...  ...  ...  ...    ...    ...    ...   ...   
760      500  0.002  ...   4.7  5.9  0.8  0.7  0.508  0.418  0.808   3.5   
761      500  0.002  ...   8.3  1.7  0.6  1.6  0.637  0.182  0.794   6.1   
762      500  0.002  ...   3.7  6.6  1.9  0.3  0.559  0.313  0.688   4.8   
763      500  0.002  ...   4.4  4.2  1.3  0.4  0.450  0.389  0.817   5.2   
764      500  0.002  ...   4.9  2.4  0.9  1.0  0.477  0.380  0.714   3.9   

     WS/48      br_id  
0    0.143  roberal01  
1    0.176  barklch01  
2    0.099  willike02  
3    0.170  wilkido01  
4    0.112  duckwke01  
..     ...        ...  
760  0.143  gilgesh01  
761  0.165  holmeri01  
762  0.128  mccontj01  
763  0.105  roziete01  
764  0.080  wiggian01  

[765 rows x 23 columns]

In [281]:
# We have to know which seasons each player was a candidate for the award
voting["cand_season"]=voting["br_id"]+voting["Year"].astype("str")
train["cand_season"]=train["br_id"]+train["season"].astype("str")

In [282]:
train["is_candidate"]=0

for idx in voting["cand_season"]:
    if idx in list(train["cand_season"]):
        train.loc[train["cand_season"]==idx,"is_candidate"]=1

In [283]:
# Players who played less than 30 games this season or the previous one shouldn't be eligible for the award
less_than_30=test[test["games"]<30]["br_id"].unique()
test=test[~test["br_id"].isin(less_than_30)]
test=test.reset_index(drop=True)

### We have to calculate the improvement of each player in the season that they were candidates for the MIP award compared to the previous season

### We create a Dataframe to calculate the differences from the previous year

In [284]:
diffs_train=pd.DataFrame(index=train[train["is_candidate"]==1].index,columns=train.columns)
diffs_test=pd.DataFrame(index=test[test["season"]==2022].index,columns=test.columns)

In [285]:
diffs_train["br_id"]=train[train["is_candidate"]==1]["br_id"]
diffs_train["team"]=train[train["is_candidate"]==1]["team"]
diffs_train["cand_season"]=train[train["is_candidate"]==1]["cand_season"]
diffs_train.drop(columns=["season","age","league","pos","is_candidate"],inplace=True)

In [286]:
diffs_test["br_id"]=test[test["season"]==2022]["br_id"]
diffs_test["team"]=test[test["season"]==2022]["team"]
diffs_test.drop(columns=["season","age","league","pos"],inplace=True)

In [287]:
train=train.reset_index(drop=True)
test=test.reset_index(drop=True)

In [288]:
for i in range(len(train.index)):
    if train.loc[i,"is_candidate"]==1:
        for k in diffs_train.columns[2:-1]:
            diffs_train.loc[i,k]=train.loc[i,k]-train.loc[i-1,k]
            #if k.endswith("%") or k.endswith("r") or k.endswith("WS") or k.endswith("BPM") or k=="WS/48" or k=="VORP" or k=="games" or k=="Games Started":
            #    diffs.loc[i,k]=data.loc[i,k]-data.loc[i-1,k]
            #else:
            #    diffs.loc[i,k]=(data.loc[i,k]/data.loc[i-1,k])-1

In [289]:
for i in range(len(test.index)):
    if test.loc[i,"season"]==2022:
        for k in diffs_test.columns[2:]:
            diffs_test.loc[i,k]=test.loc[i,k]-test.loc[i-1,k]

In [290]:
diffs_test.head(6)

br_id team games Games Started    MP MP Total   FG  FGA    FG%   3P  \
1   achiupr01  TOR    12            24  11.5      988  1.6  4.6 -0.105  0.8   
3   adamsst01  MEM    18            17  -1.4      394 -0.5 -0.2 -0.067  0.0   
5   adebaba01  MIA    -8            -8  -0.9     -318  0.2  0.5 -0.013  0.0   
7   alexani01  TOT    19             8   0.7      459 -0.3  0.5 -0.047 -0.1   
9   allengr01  MIL    16            23   2.1      546  0.4  0.3   0.03  0.3   
11  allenja01  CLE    -7            11   2.7      -55  1.9  2.0  0.059 -0.1   

    ... TOV% USG%  OWS  DWS   WS  WS/48 OBPM DBPM  BPM VORP  
1   ... -2.2 -1.0  0.1  1.1  1.2 -0.015  1.6 -0.1  1.5  0.2  
3   ...  2.1  0.3  1.5  1.3  2.8  0.044  1.4  0.9  2.3  1.3  
5   ... -0.6  1.3 -2.0  0.3 -1.6 -0.009 -1.2  0.1 -1.1 -1.0  
7   ... -1.1  0.9 -0.8  0.1 -0.6 -0.032 -0.4 -1.2 -1.6 -0.5  
9   ... -2.9 -1.6  1.3  0.2  1.5  0.009  0.8 -0.3  0.7  0.5  
11  ... -1.4  1.5  1.1  0.9  2.1  0.059  1.4  1.4  2.8  1.3  

[6 rows x 48 columns]

In [291]:
new_names=[]
for i in diffs_train.columns:
    new_names.append(i+"_diff")

diffs_train.columns=new_names

In [292]:
new_names=[]
for i in diffs_test.columns:
    new_names.append(i+"_diff")    

diffs_test.columns=new_names

In [293]:
# Merging the diff tables to the original ones
test=pd.concat([test[test["season"]==2022],diffs_test],axis=1)
train=pd.concat([train[train["is_candidate"]==1],diffs_train],axis=1)

### Preparing the model

### Classification
##### Full Data
|Dataset|Name|
|-------|----|
|X train|X_train|
|Y train|Y_train|
|Test| test|

##### Less Data (Deleted columns that might be correlated)
|Dataset|Name|
|-------|----|
|X train|X_train_del|
|Y train|Y_train|
|Test| test_del|

##### Percentile Data
|Dataset|Name|
|-------|----|
|X train|X_train_pct|
|Y train|Y_train|
|Test| test_pct|

##### Oversampled Data
|Dataset|Name|
|-------|----|
|X train|X_train_OS|
|Y train|Y_train_OS|
|Test| test|

### Regression
##### Full Data
|Dataset|Name|
|-------|----|
|X train|X_train|
|Y train|Y_train_reg|
|Test| test|

##### Less Data (Deleted columns that might be correlated)
|Dataset|Name|
|-------|----|
|X train|X_train_del|
|Y train|Y_train_reg|
|Test| test_del|

##### Scaled Data
|Dataset|Name|
|-------|----|
|X train|X_train|
|Y train|Y_train_reg_scaled|
|Test| test|

##### Percentile Data
|Dataset|Name|
|-------|----|
|X train|X_train_pct|
|Y train|Y_train_reg|
|Test| test_pct|

In [294]:
from sklearn.model_selection import cross_val_score
from sklearn.naive_bayes import GaussianNB
from sklearn.linear_model import LogisticRegression
from sklearn import tree
from sklearn.neighbors import KNeighborsClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.svm import SVC
from sklearn.ensemble import VotingClassifier
from xgboost import XGBClassifier
from sklearn.metrics import mean_squared_error
from sklearn.metrics import mean_absolute_error
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import accuracy_score
from matplotlib import pyplot as plt
from scipy. stats import skew
from sklearn.metrics import accuracy_score

### Preparing full datasets

In [295]:
# Setting player IDs as indices
test=test.set_index("br_id")
train=train.set_index("cand_season")

In [296]:
# Creating a shuffling vector
n=pd.DataFrame(train["season"].value_counts().sort_index())

fold_vec=[]
k=1
for i in n.index:
    j= n.loc[i,"season"]
    while j!= 0:
        fold_vec.append(k)
        j=j-1
    k+=1
    
fold_vec=np.array(fold_vec)

In [297]:
train=train.sort_values("season")

In [298]:
# Getting rid of columns we don't need
X_train=train.drop(columns=['br_id', 'age','team','league','is_candidate','br_id_diff','team_diff','cand_season_diff','season',
                    'pos', 'games_diff','Games Started_diff','MP Total_diff','games','Games Started','MP Total'])
test.drop(columns=['age','team','league','br_id_diff','team_diff','season','pos','games_diff',
                    'Games Started_diff','MP Total_diff','games','Games Started','MP Total'],inplace=True)

### Preparing datasets with less columns

In [299]:
# Getting rid of columns we don't need
X_train_del=X_train.drop(columns=['3PA','2PA','FTA','TRB','WS','BPM','3PA_diff','2PA_diff','FTA_diff','TRB_diff',
                    'WS_diff','BPM_diff','MP', 'FG', 'FGA', 'FG%', '3P', '3P%', '2P', '2P%', 'eFG%', 'FT', 'FT%',
                    'ORB', 'DRB', 'AST', 'STL', 'BLK', 'TOV', 'PF', 'PTS'])
test_del=test.drop(columns=['3PA','2PA','FTA','TRB','WS','BPM','3PA_diff','2PA_diff','FTA_diff','TRB_diff',
                    'WS_diff','BPM_diff','MP', 'FG', 'FGA', 'FG%', '3P', '3P%', '2P', '2P%', 'eFG%', 'FT', 'FT%',
                    'ORB', 'DRB', 'AST', 'STL', 'BLK', 'TOV', 'PF', 'PTS'])

### Preparing percentile datasets

In [300]:
# Setting up rank percentiles
train_pct_rank= []

for i in range(1,37):
    l=len(X_train[fold_vec==i].rank(pct=True))
    for j in range(l):
        row=X_train[fold_vec==i].rank(pct=True).iloc[j]
        train_pct_rank.append([item for item in row if item])

test_pct=test.rank(pct=True)

In [301]:
train_pct_rank=pd.DataFrame(train_pct_rank,index=X_train.index,columns=X_train.columns)

### Preparing Regression datasets

In [302]:
# Getting the voting share
votes=voting[["cand_season","Share"]]

In [303]:
train_reg=X_train
train_reg=train_reg.merge(votes,left_on="cand_season",right_on="cand_season",how="inner")

train_reg=train_reg.set_index("cand_season")

In [304]:
#train_reg=train_reg.sort_values("season")

In [305]:
Y_train_reg=train_reg["Share"]

### Scaled voting data for Regression

In [306]:
from sklearn.linear_model import LinearRegression
from xgboost.sklearn import XGBRegressor
from sklearn.linear_model import SGDRegressor
from sklearn.svm import SVR
from sklearn.tree import DecisionTreeRegressor
from sklearn.ensemble import RandomForestRegressor
from sklearn.preprocessing import MinMaxScaler

In [307]:
# Scaling voting shares
scaled_total=[]
for i in range(1,36):
    scaler = MinMaxScaler()
    to_scale=Y_train_reg[fold_vec==i]
    scaled = scaler.fit_transform(to_scale.values.reshape(-1, 1))
    for j in range(len(scaled)):
        scaled_total.append(scaled[j])
        
Y_train_reg_scaled=pd.DataFrame(scaled_total)
Y_train_reg_scaled=Y_train_reg_scaled[0]

### Adding the winners for the Classification datasets

In [308]:
# Finding the winners
voting["winner"]=np.where(voting["Rank"]=="1", 1, 0)
winners=voting[["cand_season","winner"]]

In [309]:
X_train=X_train.merge(winners,left_on="cand_season",right_on="cand_season",how="inner")
train_pct_rank=train_pct_rank.merge(winners,left_on="cand_season",right_on="cand_season",how="inner")

X_train=X_train.set_index("cand_season")
train_pct_rank=train_pct_rank.set_index("cand_season")

### Preparing OverSampled datasets

In [310]:
train_OS=pd.DataFrame()

for i in range(1,36):
    to_check=X_train.iloc[fold_vec==i]
    l=len(to_check)
    to_repeat=to_check[to_check["winner"]==1]
    for j in to_check.index:
        row=X_train.filter(like=j,axis=0)
        if row.iloc[:,-1][0]==1:
            for k in range(l-1):
                train_OS=train_OS.append(row)
        else:
            train_OS=train_OS.append(row)

In [311]:
fold_vec_OS=[]

for i in range(1,36):
    for j in range((sum(fold_vec==i)-1)*2):
        fold_vec_OS.append(i)
        
fold_vec_OS=np.array(fold_vec_OS)

In [312]:
X_train_OS=train_OS.drop(columns="winner")
Y_train_OS=train_OS["winner"]

In [313]:
X_train_OS[X_train_OS.columns]=X_train_OS[X_train_OS.columns].apply(pd.to_numeric, errors='coerce')

### Finishing up the datasets

In [314]:
# Converting values to numerics
X_train[X_train.columns]=X_train[X_train.columns].apply(pd.to_numeric, errors='coerce')
X_train_del[X_train_del.columns]=X_train_del[X_train_del.columns].apply(pd.to_numeric, errors='coerce')
test[test.columns]=test[test.columns].apply(pd.to_numeric, errors='coerce')

In [315]:
Y_train=X_train["winner"]
X_train=X_train.drop(columns="winner")

In [316]:
Y_train_pct=train_pct_rank["winner"]
X_train_pct=train_pct_rank.drop(columns="winner")

### Creating functions to evaluate and predict

In [318]:
# Create CV function
def cv(x_train_data,y_train_data, test_data, model):
    out = np.zeros(35)
    acc=0
    
    # Calculating accuracy
    for i in range(1,36):
        test_i = fold_vec == i
        train_i = fold_vec != i
        x_test = x_train_data.iloc[test_i, :]
        x_train = x_train_data.iloc[train_i, :]
        y_test= y_train_data[test_i]
        y_train = y_train_data[train_i]
        mod = model
        mod.fit(x_train, y_train)
        prob = mod.predict_proba(x_test)
        prob = prob[:, 1]
        max=prob.max()
        pred=prob==max
        if y_test[pred][0]==1:
            acc+=1
        out[i-1]=accuracy_score(y_test,pred)
        
    print("Average Accuracy: ",np.mean(out))
    #print("Accuracy: ",acc,"/35")
    #print("Accuracy%: ",acc/35)
    
    # Making predictions
    mod=model
    mod.fit(x_train_data, y_train_data)
    mip_prob=mod.predict_proba(test_data)
    mip_prob = mip_prob[:, 1]
    mip = {'br_id': test.index, 'pred': mip_prob}
    mip = pd.DataFrame(data=mip).sort_values("pred",ascending=False).head()
    #mip=mip[mip["pred"]!=1]
    return mip

In [319]:
# Create CV function - Oversampling
def cv_OS(x_train_data,y_train_data, test_data, model):
    out = np.zeros(35)
    acc=0
    
    # Calculating accuracy
    for i in range(1,36):
        test_i = fold_vec_OS == i
        train_i = fold_vec_OS != i
        x_test = x_train_data.iloc[test_i, :]
        x_train = x_train_data.iloc[train_i, :]
        y_test= y_train_data[test_i]
        y_train = y_train_data[train_i]
        mod = model
        mod.fit(x_train, y_train)
        prob = mod.predict_proba(x_test)
        prob = prob[:, 1]
        max=prob.max()
        pred=prob==max
        if y_test[pred][0]==1:
            acc+=1
        out[i-1]=accuracy_score(y_test,pred)
        
    print("Average Accuracy: ",np.mean(out))
    #print("Accuracy: ",acc,"/35")
    #print("Accuracy%: ",acc/35)
    
    # Making predictions
    mod=model
    mod.fit(x_train_data, y_train_data)
    mip_prob=mod.predict_proba(test_data)
    mip_prob = mip_prob[:, 1]
    mip = {'br_id': test.index, 'pred': mip_prob}
    mip = pd.DataFrame(data=mip).sort_values("pred",ascending=False).head()
    #mip=mip[mip["pred"]!=1]
    return mip

In [320]:
# Create CV function - Regression
def cv_reg(x_train_data,y_train_data, test_data, model):
    out = np.zeros(35)
    acc=0
    
    # Calculating accuracy
    for i in range(1,36):
        test_i = fold_vec == i
        train_i = fold_vec != i
        x_test = x_train_data.iloc[test_i, :]
        x_train = x_train_data.iloc[train_i, :]
        y_test= y_train_data[test_i]
        y_train = y_train_data[train_i]
        mod = model
        mod.fit(x_train, y_train)
        prob = mod.predict(x_test)
        max=prob.max()
        pred=prob==max
        #print(y_test)
        if y_test[pred].index[0]==y_test.sort_values(ascending=False).index[0]:
            acc+=1
            #print("GOT HIM!")
            #print(y_test[pred].index[0])
        out[i-1]=mean_absolute_error(y_test,prob)
        
    print("Average MAE: ",np.mean(out))
    #print("Accuracy: ",acc,"/35")
    #print("Accuracy%: ",acc/35)
    
    # Making predictions
    mod=model
    mod.fit(x_train_data, y_train_data)
    mip_prob=mod.predict(test_data)
    mip = {'br_id': test.index, 'pred': mip_prob}
    mip = pd.DataFrame(data=mip).sort_values("pred",ascending=False).head()
    return mip

### Running models

#### Classification - Full Data

In [321]:
# Logistic Regression
cv(X_train,Y_train,test,LogisticRegression(max_iter=9000))

Average Accuracy:  0.9007357183569719


br_id      pred
19    banede01  0.213954
100  garlada01  0.090322
16    ballla01  0.090196
238  poolejo01  0.088311
206  moranja01  0.074884

In [322]:
# Gaussian Naive Bayes
cv(X_train,Y_train,test,GaussianNB())

Average Accuracy:  0.9272322660305463


br_id      pred
209  murrade01  1.000000
100  garlada01  0.999674
191  maxeyty01  0.999227
206  moranja01  0.999205
19    banede01  0.962836

In [323]:
# Decision Tree
cv(X_train,Y_train,test,tree.DecisionTreeClassifier())

Average Accuracy:  0.6490369652634211


br_id  pred
77   derozde01   1.0
236  poeltja01   1.0
150  jamesle01   1.0
238  poolejo01   1.0
0    achiupr01   0.0

In [324]:
# KNN
cv(X_train,Y_train,test,KNeighborsClassifier(n_neighbors=200))

Average Accuracy:  0.9045671996096003


br_id   pred
206  moranja01  0.095
209  murrade01  0.085
100  garlada01  0.080
81   doncilu01  0.075
309  youngtr01  0.075

In [325]:
# Random Forest
cv(X_train,Y_train,test,RandomForestClassifier())

Average Accuracy:  0.9152395586027477


br_id  pred
90   favorde01  0.17
77   derozde01  0.16
105  goberru01  0.16
48   brownmo01  0.15
263  simonan01  0.14

In [326]:
# Support Vector Classifier
cv(X_train,Y_train,test,SVC(probability = True))

Average Accuracy:  0.9130590670512883


br_id      pred
209  murrade01  0.111708
16    ballla01  0.084831
238  poolejo01  0.080598
43   bridgmi02  0.078544
110  greendr01  0.078122

In [327]:
# Xtreme Gradient Boosting
cv(X_train,Y_train,test,XGBClassifier(random_state =1,eval_metric="logloss",use_label_encoder=False))

Average Accuracy:  0.9272299712305353


br_id      pred
238  poolejo01  0.244673
263  simonan01  0.033618
209  murrade01  0.030788
206  moranja01  0.024568
19    banede01  0.019154

#### Classification - Percentile Data

In [328]:
# Logistic Regression
cv(X_train_pct,Y_train,test_pct,LogisticRegression(max_iter=9000))

Average Accuracy:  0.9112797189484108


br_id      pred
100  garlada01  0.864529
238  poolejo01  0.851643
19    banede01  0.836938
16    ballla01  0.821293
131  herroty01  0.779366

In [329]:
# Gaussian Naive Bayes
cv(X_train_pct,Y_train,test_pct,GaussianNB())

Average Accuracy:  0.9146153447347755


br_id  pred
100  garlada01   1.0
209  murrade01   1.0
19    banede01   1.0
206  moranja01   1.0
238  poolejo01   1.0

In [330]:
# Decision Tree
cv(X_train_pct,Y_train,test_pct,tree.DecisionTreeClassifier())

Average Accuracy:  0.6500980404725913


br_id  pred
0    achiupr01   1.0
282  tuckepj01   1.0
47   brownja02   1.0
267  stevela01   1.0
58   capelca01   1.0

In [331]:
# KNN
cv(X_train_pct,Y_train,test_pct,KNeighborsClassifier(n_neighbors=200))

Average Accuracy:  0.9069191816886032


br_id   pred
77   derozde01  0.130
206  moranja01  0.130
238  poolejo01  0.130
19    banede01  0.130
100  garlada01  0.125

In [332]:
# Random Forest
cv(X_train_pct,Y_train,test_pct,RandomForestClassifier())

Average Accuracy:  0.9217750276493005


br_id  pred
19    banede01  0.41
191  maxeyty01  0.40
263  simonan01  0.40
209  murrade01  0.39
238  poolejo01  0.37

In [333]:
# Support Vector Classifier
cv(X_train_pct,Y_train,test_pct,SVC(probability = True))

Average Accuracy:  0.9052019241941455


br_id      pred
238  poolejo01  0.636500
19    banede01  0.618177
131  herroty01  0.481723
16    ballla01  0.467636
100  garlada01  0.446757

In [334]:
# Xtreme Gradient Boosting
cv(X_train_pct,Y_train,test_pct,XGBClassifier(random_state =1,eval_metric="logloss",use_label_encoder=False))

Average Accuracy:  0.9009739415459341


br_id      pred
131  herroty01  0.973719
117  halibty01  0.923318
77   derozde01  0.921346
191  maxeyty01  0.914753
100  garlada01  0.882222

#### Classification - Oversampled data

In [335]:
# Logistic Regression
cv_OS(X_train_OS,Y_train_OS,test,LogisticRegression(max_iter=9000))

Average Accuracy:  0.5889670949961768


br_id      pred
238  poolejo01  0.942478
19    banede01  0.876328
16    ballla01  0.831462
195  mcgeeja01  0.724263
191  maxeyty01  0.681431

In [336]:
# Gaussian Naive Bayes
cv_OS(X_train_OS,Y_train_OS,test,GaussianNB())

Average Accuracy:  0.6529990395949481


br_id      pred
209  murrade01  1.000000
100  garlada01  0.999998
206  moranja01  0.999993
191  maxeyty01  0.999985
19    banede01  0.999889

In [337]:
# Decision Tree
cv_OS(X_train_OS,Y_train_OS,test,tree.DecisionTreeClassifier())

Average Accuracy:  0.524569796669451


br_id  pred
0    achiupr01   0.0
205   monkma01   0.0
212  nancela02   0.0
211  mykhasv01   0.0
210  muscami01   0.0

In [338]:
# KNN
cv_OS(X_train_OS,Y_train_OS,test,KNeighborsClassifier(n_neighbors=200))

Average Accuracy:  0.576316039415404


br_id   pred
238  poolejo01  0.810
209  murrade01  0.775
16    ballla01  0.725
206  moranja01  0.715
43   bridgmi02  0.710

In [339]:
# Random Forest
cv_OS(X_train_OS,Y_train_OS,test,RandomForestClassifier())

Average Accuracy:  0.6194496198270408


br_id  pred
209  murrade01  0.47
206  moranja01  0.35
191  maxeyty01  0.31
238  poolejo01  0.21
43   bridgmi02  0.16

In [340]:
# Support Vector Classifier
cv_OS(X_train_OS,Y_train_OS,test,SVC(probability = True))

Average Accuracy:  0.6376239869898381


br_id      pred
43   bridgmi02  0.860071
19    banede01  0.851241
206  moranja01  0.750960
16    ballla01  0.673429
131  herroty01  0.645821

In [341]:
# Xtreme Gradient Boosting
cv_OS(X_train_OS,Y_train_OS,test,XGBClassifier(random_state =1,eval_metric="logloss",use_label_encoder=False))

Average Accuracy:  0.5766721924789303


br_id      pred
209  murrade01  0.736557
206  moranja01  0.149616
263  simonan01  0.066439
191  maxeyty01  0.025697
150  jamesle01  0.012842

#### Regression - Full Data

In [342]:
# Linear Regression
cv_reg(X_train,Y_train_reg,test,LinearRegression())

Average MAE:  0.08535045602755965


br_id      pred
206  moranja01  0.173358
209  murrade01  0.170128
19    banede01  0.168486
100  garlada01  0.167133
191  maxeyty01  0.156689

In [343]:
# XGBRegressor
cv_reg(X_train,Y_train_reg,test,XGBRegressor())

Average MAE:  0.07777807915028083


br_id      pred
263  simonan01  0.375319
238  poolejo01  0.331430
209  murrade01  0.331318
191  maxeyty01  0.315397
19    banede01  0.299530

In [344]:
# SGDRegressor
cv_reg(X_train,Y_train_reg,test,SGDRegressor())

Average MAE:  859853783183.1876


br_id          pred
206  moranja01  1.927196e+12
149  jacksre01  1.785607e+12
102  georgpa01  1.764653e+12
209  murrade01  1.764352e+12
84   duranke01  1.666805e+12

In [345]:
# SVR
cv_reg(X_train,Y_train_reg,test,SVR())

Average MAE:  0.08310228296596765


br_id      pred
209  murrade01  0.185296
19    banede01  0.150957
43   bridgmi02  0.150793
100  garlada01  0.138847
206  moranja01  0.135138

In [346]:
# Decision Tree Regressor
cv_reg(X_train,Y_train_reg,test,DecisionTreeRegressor())

Average MAE:  0.0881382355060838


br_id   pred
159  jonestr01  0.442
256  rubiori01  0.243
150  jamesle01  0.204
299  whiteha01  0.196
195  mcgeeja01  0.196

In [347]:
# Random Forest Regressor
cv_reg(X_train,Y_train_reg,test,RandomForestRegressor())

Average MAE:  0.07380532596689499


br_id     pred
206  moranja01  0.21187
19    banede01  0.20379
191  maxeyty01  0.17470
263  simonan01  0.15067
48   brownmo01  0.14167

#### Regression - Scaled data

In [348]:
# Linear Regression
cv_reg(X_train,Y_train_reg_scaled,test,LinearRegression())

Average MAE:  0.1618541002407213


br_id      pred
191  maxeyty01  0.278783
238  poolejo01  0.274231
100  garlada01  0.256954
19    banede01  0.239913
209  murrade01  0.218565

In [349]:
# XGBRegressor
cv_reg(X_train,Y_train_reg_scaled,test,XGBRegressor())

Average MAE:  0.14303175531465925


br_id      pred
191  maxeyty01  0.620885
19    banede01  0.575262
209  murrade01  0.484171
284  valanjo01  0.446163
206  moranja01  0.391554

In [350]:
# SGDRegressor
cv_reg(X_train,Y_train_reg_scaled,test,SGDRegressor())

Average MAE:  1035013699192.8817


br_id          pred
150  jamesle01  3.238953e+12
81   doncilu01  3.137745e+12
296  westbru01  3.109551e+12
246  randlju01  3.074435e+12
289  vucevni01  2.945059e+12

In [351]:
# SVR
cv_reg(X_train,Y_train_reg_scaled,test,SVR())

Average MAE:  0.13756637893985155


br_id      pred
209  murrade01  0.199649
19    banede01  0.192039
191  maxeyty01  0.176236
43   bridgmi02  0.168141
234  paytoel01  0.156876

In [352]:
# Decision Tree Regressor
cv_reg(X_train,Y_train_reg_scaled,test,DecisionTreeRegressor())

Average MAE:  0.17490410202304477


br_id  pred
90   favorde01   1.0
238  poolejo01   1.0
159  jonestr01   1.0
190  mathega01   1.0
19    banede01   1.0

In [353]:
# Random Forest Regressor
cv_reg(X_train,Y_train_reg_scaled,test,RandomForestRegressor())

Average MAE:  0.1404931542736897


br_id      pred
191  maxeyty01  0.371623
19    banede01  0.366869
90   favorde01  0.328866
209  murrade01  0.320944
206  moranja01  0.252240

#### Regression - Percentile Data

In [354]:
# Linear Regression
cv_reg(X_train_pct,Y_train_reg,test_pct,LinearRegression())

Average MAE:  0.08569286672155617


br_id      pred
100  garlada01  0.316074
16    ballla01  0.312577
209  murrade01  0.306020
282  tuckepj01  0.292919
126  harteis01  0.271061

In [355]:
# XGBRegressor
cv_reg(X_train_pct,Y_train_reg,test_pct,XGBRegressor())

Average MAE:  0.07680790371884436


br_id      pred
19    banede01  0.673454
7    antetgi01  0.610636
238  poolejo01  0.523324
206  moranja01  0.472395
209  murrade01  0.471159

In [356]:
# SGDRegressor
cv_reg(X_train_pct,Y_train_reg,test_pct,SGDRegressor())

Average MAE:  0.07726506141106565


br_id      pred
43   bridgmi02  0.180949
131  herroty01  0.172278
191  maxeyty01  0.171088
209  murrade01  0.169584
100  garlada01  0.169551

In [357]:
# SVR
cv_reg(X_train_pct,Y_train_reg,test_pct,SVR())

Average MAE:  0.0921687675882535


br_id      pred
19    banede01  0.546734
238  poolejo01  0.522004
209  murrade01  0.516794
16    ballla01  0.514389
131  herroty01  0.486947

In [358]:
# Decision Tree Regressor
cv_reg(X_train_pct,Y_train_reg,test_pct,DecisionTreeRegressor())

Average MAE:  0.09071845875047999


br_id   pred
46   brownbr01  0.986
158  jonesda03  0.986
126  harteis01  0.986
114  greenjo02  0.986
88   embiijo01  0.986

In [359]:
# Random Forest Regressor
cv_reg(X_train_pct,Y_train_reg,test_pct,RandomForestRegressor())

Average MAE:  0.0736005895901705


br_id     pred
7    antetgi01  0.42011
43   bridgmi02  0.38368
77   derozde01  0.36500
125   hartjo01  0.36188
191  maxeyty01  0.36084

In [360]:
# Cross Validation - Classification
# Full dataset
cv_lr_full=np.mean(cross_val_score(LogisticRegression(max_iter=9000), X_train, Y_train,scoring='accuracy'))
cv_knn_full=np.mean(cross_val_score(KNeighborsClassifier(n_neighbors=200), X_train, Y_train,scoring='accuracy'))
cv_rf_full=np.mean(cross_val_score(RandomForestClassifier(), X_train, Y_train,scoring='accuracy'))
cv_xgb_full=np.mean(cross_val_score(XGBClassifier(random_state =1,eval_metric="logloss",use_label_encoder=False), X_train, Y_train,scoring='accuracy'))

cv_full=(cv_lr_full+cv_knn_full+cv_rf_full+cv_xgb_full)/4

# Percentile ranked dataset
cv_lr_pct=np.mean(cross_val_score(LogisticRegression(max_iter=9000), X_train_pct, Y_train,scoring='accuracy'))
cv_knn_pct=np.mean(cross_val_score(KNeighborsClassifier(n_neighbors=200), X_train_pct, Y_train,scoring='accuracy'))
cv_rf_pct=np.mean(cross_val_score(RandomForestClassifier(), X_train_pct, Y_train,scoring='accuracy'))
cv_xgb_pct=np.mean(cross_val_score(XGBClassifier(random_state =1,eval_metric="logloss",use_label_encoder=False), X_train_pct, Y_train,scoring='accuracy'))

cv_pct=(cv_lr_pct+cv_knn_pct+cv_rf_pct+cv_xgb_pct)/4

# Oversampled dataset
cv_lr_OS=np.mean(cross_val_score(LogisticRegression(max_iter=9000), X_train_OS, Y_train_OS,scoring='accuracy'))
cv_knn_OS=np.mean(cross_val_score(KNeighborsClassifier(n_neighbors=200), X_train_OS, Y_train_OS,scoring='accuracy'))
cv_rf_OS=np.mean(cross_val_score(RandomForestClassifier(), X_train_OS, Y_train_OS,scoring='accuracy'))
cv_xgb_OS=np.mean(cross_val_score(XGBClassifier(random_state =1,eval_metric="logloss",use_label_encoder=False), X_train_OS, Y_train_OS,scoring='accuracy'))

cv_OS=(cv_lr_OS+cv_knn_OS+cv_rf_OS+cv_xgb_OS)/4

In [361]:
print("Full dataset average accuracy: ",cv_full)
print("Percentile ranked average accuracy: ",cv_pct)
print("Oversampled dataset average accuracy: ",cv_OS)

Full dataset average accuracy:  0.9459175670965494
Percentile ranked average accuracy:  0.9528407110491461
Oversampled dataset average accuracy:  0.6352720439088415


### It turns out that the Percentile ranked dataset gives us the best results

|Dataset|Average Accuracy|
|-------|----------------|
|Full|94.62%|
|Percentile Ranked|**95.32%**|
|Oversampled|63.61%|

### We get the best results by using the percentile datasets. We will now see which of our variables are the most important

### Feature selection for classification
#### We will use the ANOVA f-test

In [362]:
from sklearn.feature_selection import SelectKBest
from sklearn.feature_selection import f_classif
from matplotlib import pyplot

In [363]:
fs = SelectKBest(score_func=f_classif, k=20)
fs.fit(X_train_pct, Y_train)

SelectKBest(k=20)

In [364]:
scores=pd.DataFrame(fs.scores_,index=X_train_pct.columns,columns=["Score"])

In [365]:
scores.sort_values("Score",ascending=False).head(15)

Score
PTS_diff   34.994636
FT_diff    32.948742
FG_diff    25.614086
FTA_diff   22.430626
2P_diff    21.325304
2PA_diff   20.097137
FGA_diff   20.037780
TOV        18.655064
WS_diff    17.945469
VORP_diff  17.926252
OWS_diff   17.381584
OBPM_diff  17.204031
TOV_diff   16.733029
PTS        16.308175
PER_diff   14.787585

In [366]:
fs = SelectKBest(score_func=f_classif, k=15)
fs.fit(X_train_pct, Y_train)
scores=pd.DataFrame(fs.scores_,index=X_train_pct.columns,columns=["Score"])

In [367]:
X_train_pct_clf_fs=X_train_pct.iloc[:,fs.get_support(indices=True)]
test_pct_clf_fs=test.iloc[:,fs.get_support(indices=True)]

#### According to the ANOVA f-test, the 10 most important variables are: PTS_diff, FT_diff, FG_diff, FTA_diff, 2P_diff, 2PA_diff, FGA_diff, TOV, VORP_diff and WS_diff

### Let's run the models again for this dataset

In [368]:
# Logistic Regression
cv(X_train_pct_clf_fs,Y_train,test_pct_clf_fs,LogisticRegression(max_iter=9000))

Average Accuracy:  0.9271153447347755


br_id  pred
100  garlada01   1.0
88   embiijo01   1.0
309  youngtr01   1.0
43   bridgmi02   1.0
209  murrade01   1.0

In [369]:
# KNN
cv(X_train_pct_clf_fs,Y_train,test_pct_clf_fs,KNeighborsClassifier(n_neighbors=100))

Average Accuracy:  0.9218001835201227


br_id  pred
191  maxeyty01  0.20
65   coffeam01  0.19
19    banede01  0.19
278  toppiob01  0.19
226  okongon01  0.18

In [370]:
# Random Forest
cv(X_train_pct_clf_fs,Y_train,test_pct_clf_fs,RandomForestClassifier())

Average Accuracy:  0.9165999473532522


br_id  pred
278  toppiob01  0.58
43   bridgmi02  0.55
206  moranja01  0.55
125   hartjo01  0.54
191  maxeyty01  0.54

In [371]:
# Xtreme Gradient Boosting
cv(X_train_pct_clf_fs,Y_train,test_pct_clf_fs,XGBClassifier(random_state =1,eval_metric="logloss",use_label_encoder=False))

Average Accuracy:  0.9135773488912768


br_id      pred
209  murrade01  0.757892
278  toppiob01  0.734844
125   hartjo01  0.707824
19    banede01  0.702739
43   bridgmi02  0.702739

In [372]:
# Cross Validation - Regression
# Full dataset
cv_linr_full=np.mean(cross_val_score(LinearRegression(), X_train, Y_train_reg,scoring='neg_mean_absolute_error'))
cv_svr_full=np.mean(cross_val_score(SVR(), X_train, Y_train_reg,scoring='neg_mean_absolute_error'))
cv_rfreg_full=np.mean(cross_val_score(RandomForestRegressor(), X_train, Y_train_reg,scoring='neg_mean_absolute_error'))

cv_full_reg=(cv_linr_full+cv_svr_full+cv_rfreg_full)/3

# Percentile ranked dataset
cv_linr_pct=np.mean(cross_val_score(LinearRegression(), X_train_pct,Y_train_reg,scoring='neg_mean_absolute_error'))
cv_svr_pct=np.mean(cross_val_score(SVR(), X_train_pct,Y_train_reg,scoring='neg_mean_absolute_error'))
cv_rfreg_pct=np.mean(cross_val_score(RandomForestRegressor(), X_train_pct,Y_train_reg,scoring='neg_mean_absolute_error'))

cv_pct_reg=(cv_linr_pct+cv_svr_pct+cv_rfreg_pct)/3

# Scaled dataset
cv_linr_scaled=np.mean(cross_val_score(LinearRegression(), X_train,Y_train_reg_scaled,scoring='neg_mean_absolute_error'))
cv_svr_scaled=np.mean(cross_val_score(SVR(), X_train,Y_train_reg_scaled,scoring='neg_mean_absolute_error'))
cv_rfreg_scaled=np.mean(cross_val_score(RandomForestRegressor(), X_train,Y_train_reg_scaled,scoring='neg_mean_absolute_error'))

cv_scaled_reg=(cv_linr_scaled+cv_svr_scaled+cv_rfreg_scaled)/3

In [373]:
print("Full dataset average Neg. Mean Absolute Error: ",cv_full_reg)
print("Percentile ranked dataset average Neg. Mean Absolute Error: ",cv_pct_reg)
print("Scaled dataset average Neg. Mean Absolute Error: ",cv_scaled_reg)

Full dataset average Neg. Mean Absolute Error:  -0.08135222726008058
Percentile ranked dataset average Neg. Mean Absolute Error:  -0.08520256697349438
Scaled dataset average Neg. Mean Absolute Error:  -0.13738219658819942


### It turns out that the Full dataset gives us the best results

|Dataset|Average Accuracy|
|-------|----------------|
|Full|**-0.082**|
|Percentile Ranked|-0.085|
|Scaled|-0.137|

### Feature Selection for Regression
#### We will use Pearson's correlation

In [374]:
corr=pd.concat([X_train,Y_train_reg],axis=1).corr()["Share"]

In [375]:
abs_corr=abs(corr).sort_values(ascending=False)

In [376]:
pd.DataFrame(abs_corr[1:16])

Share
PTS_diff   0.321980
FG_diff    0.296134
FT_diff    0.295700
VORP_diff  0.295206
2P_diff    0.268514
OWS_diff   0.266211
FTA_diff   0.264592
FGA_diff   0.264246
2PA_diff   0.254101
WS_diff    0.240931
TOV_diff   0.219859
AST_diff   0.211251
TOV        0.193195
MP         0.184819
PTS        0.182933

#### According to Pearson's correlation, the 10 most important variables are: PTS_diff, FG_diff, FT_diff, VORP_diff, 2P_diff, OWS_diff, FTA_diff, FGA_diff, 2PA_diff and WS_diff

In [377]:
X_train_reg_fs=X_train[abs_corr[1:16].index]
test_reg_fs=test[abs_corr[1:16].index]

### Running the models again

In [378]:
# Linear Regression
cv_reg(X_train_reg_fs,Y_train_reg,test_reg_fs,LinearRegression())

Average MAE:  0.07780793324268578


br_id      pred
206  moranja01  0.172777
191  maxeyty01  0.131956
100  garlada01  0.127889
209  murrade01  0.119554
19    banede01  0.108491

In [379]:
# SVR
cv_reg(X_train_reg_fs,Y_train_reg,test_reg_fs,SVR())

Average MAE:  0.09538104752229537


br_id      pred
191  maxeyty01  0.141415
43   bridgmi02  0.126864
209  murrade01  0.123917
19    banede01  0.121620
206  moranja01  0.119861

In [380]:
# Random Forest Regressor
cv_reg(X_train_reg_fs,Y_train_reg,test_reg_fs,RandomForestRegressor())

Average MAE:  0.07416896725977322


br_id     pred
191  maxeyty01  0.23833
209  murrade01  0.22505
19    banede01  0.16365
150  jamesle01  0.15266
206  moranja01  0.14514

### Voting

In [381]:
from sklearn.ensemble import VotingClassifier
from sklearn.ensemble import VotingRegressor

### For the classification

In [382]:
voting_clf_soft = VotingClassifier(estimators = 
                                   [('lf', LogisticRegression()),
                                    ('knn',KNeighborsClassifier(n_neighbors=100)),
                                    ('rf',RandomForestClassifier()),
                                    ('xgb',XGBClassifier(random_state =1,eval_metric="logloss",use_label_encoder=False))]
                                   , voting = 'soft') 

In [383]:
cv(X_train_pct_clf_fs,Y_train,test_pct_clf_fs,voting_clf_soft)

Average Accuracy:  0.9119539722679002


br_id      pred
209  murrade01  0.634473
278  toppiob01  0.626211
191  maxeyty01  0.620685
19    banede01  0.615685
131  herroty01  0.610685

### For the regression

In [384]:
voting_reg=VotingRegressor(estimators = 
                           [('lr', LinearRegression()),
                           ('svr',SVR()),
                           ('rf',RandomForestRegressor())]) 

In [385]:
cv_reg(X_train_reg_fs,Y_train_reg,test_reg_fs,voting_reg)

Average MAE:  0.07782307105706687


br_id      pred
191  maxeyty01  0.171550
206  moranja01  0.151109
209  murrade01  0.143950
19    banede01  0.132207
100  garlada01  0.110332